runtime/calamari/src/migrations/asset_id.rs

```diff
@@ -0,0 +1,680 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Manta.  If not, see <http://www.gnu.org/licenses/>.
+
+#![allow(clippy::unnecessary_cast)]
+
+use crate::sp_api_hidden_includes_construct_runtime::hidden_include::StorageHasher;
+use codec::{Decode, Encode};
+use core::marker::PhantomData;
+#[allow(deprecated)]
+use frame_support::migration::{
+    get_storage_value, put_storage_value, storage_key_iter, take_storage_value,
+};
+use frame_support::{
+    dispatch::GetStorageVersion,
+    pallet_prelude::Weight,
+    storage_alias,
+    traits::{Currency, Get, OnRuntimeUpgrade, PalletInfoAccess, StorageVersion},
+    Blake2_128Concat,
+};
+use manta_primitives::{
+    assets::{AssetLocation, AssetRegistryMetadata, AssetStorageMetadata},
+    types::Balance,
+};
+use scale_info::TypeInfo;
+use sp_core::H160;
+use sp_runtime::BoundedVec;
+use sp_std::vec::Vec;
+
+pub type DepositBalanceOf<T, I = ()> = <<T as pallet_assets::Config<I>>::Currency as Currency<
+    <T as frame_system::Config>::AccountId,
+>>::Balance;
+pub type AssetAccountOf<T, I> = pallet_assets::AssetAccount<
+    <T as pallet_assets::Config<I>>::Balance,
+    DepositBalanceOf<T, I>,
+    <T as pallet_assets::Config<I>>::Extra,
+>;
+
+type OldAssetId = u32;
+type NewAssetId = u128;
+
+pub mod old {
+    use crate::migrations::asset_id::OldAssetId;
+    use frame_support::{storage_alias, Blake2_128Concat};
+
+    #[storage_alias]
+    pub type Account<T: frame_system::Config<I>, I: 'static = ()> = StorageDoubleMap<
+        Assets,
+        Blake2_128Concat,
+        OldAssetId,
+        Blake2_128Concat,
+        <T as frame_system::Config>::AccountId,
+        super::AssetAccountOf<T, I>,
+    >;
+}
+
+#[storage_alias]
+type Account<T: frame_system::Config<I>, I: 'static = ()> = StorageDoubleMap<
+    Assets,
+    Blake2_128Concat,
+    NewAssetId,
+    Blake2_128Concat,
+    <T as frame_system::Config>::AccountId,
+    AssetAccountOf<T, I>,
+>;
+
+type AssetMapKVP<T> = (
+    OldAssetId,
+    pallet_assets::AssetDetails<
+        <T as pallet_assets::Config>::Balance,
+        <T as frame_system::Config>::AccountId,
+        DepositBalanceOf<T>,
+    >,
+);
+
+type MetadataMapKVP<T> = (
+    OldAssetId,
+    pallet_assets::AssetMetadata<
+        DepositBalanceOf<T>,
+        BoundedVec<u8, <T as pallet_assets::Config>::StringLimit>,
+    >,
+);
+
+#[derive(Clone, Debug, Decode, Encode, Eq, Hash, Ord, PartialEq, PartialOrd, TypeInfo)]
+pub struct OldAssetRegistrarMetadata {
+    pub name: Vec<u8>,
+    pub symbol: Vec<u8>,
+    pub decimals: u8,
+    pub evm_address: Option<H160>,
+    pub is_frozen: bool,
+    pub min_balance: Balance,
+    pub is_sufficient: bool,
+}
+
+pub struct AssetIdMigration<T, M, A>(PhantomData<T>, PhantomData<M>, PhantomData<A>);
```

Comment:
could you document the meanings of TMA here? the trait bounds aren't enough to infer them

Reply:
Actually i just need the T, will get rid of the rest, and that should clear things up.